lasagne/layers/input.py

```diff
@@ -67,6 +67,9 @@ def __init__(self, shape, input_var=None, name=None, **kwargs):
             if input_var.ndim != ndim:
                 raise ValueError("shape has %d dimensions, but variable has "
                                  "%d" % (ndim, input_var.ndim))
+            broad_pattern = [s == 1 for s in self.shape]
+            input_var = T.patternbroadcast(input_var, broad_pattern)
```

Comment:
> You mean to say they both referred to the same object before(Prior to this PR) and now, they are different objects ?

Exactly.

> As in, if I understand the concern correctly, this does a deep copying whereas something like shallow copy is desired?

Hmm, no. It's just that it changes behaviour people might rely on.

```python
def build_model(input_var1, input_var2):
    net = something very complicated
    return net

input_var1, input_var2 = T.matrices('ab')
net = build_model(input_var1, input_var2)

# find layer associated with input_var1
input_layer1 = next(layer for layer in lasagne.layers.get_all_layers(network) if layer.input_var is input_var1)

# compute expression replacing it with something else
something = ...
output = lasagne.layers.get_output(net, {input_layer1: something})
```
This would work before, but not after this PR. I don't know if and how many users rely on this, but in my mental model, the input variable of an input layer is always what I constructed it with.